PICMI Species handling refactoring

chillenzer opened the pull request:
# Species Refactoring
This PR is a complete rewrite of our `picmi.Species` class and the related protocols. It is somewhat related to #5500, although not quite aligned with the original vision of how that's going to proceed. This PR removes PyPIConGPU as an officially supported interface and removes all pypicongpu unit tests.

## New Species protocol
The previous protocol of how species were handled in PyPIConGPU/PICMI had several significant drawbacks. The two main ones are:
- It was based on a stateful `InitManager` instance that could not be re-used after having been "baked" once.
- It was implemented on the PyPIConGPU level while it semantically belongs into our implementation of the PICMI level.

The main problem that this construction tried address is the fact that an inversion of some dependencies happens from the PICMI level to the PyPIConGPU level. For example, on the PICMI level the `Species` contains the `initial_distribution` but on the PyPIConGPU level we have a `SimpleDensity` operation that contains a (list of) `Species`. Also, various entities in the code impose requirements onto the `Species` that it must be informed about.

The new species protocol is as follows:
- The `picmi.Species` provides a `register_requirements` public method. Consumers of the species are allowed to call this before any `get_as_pypicongpu()` calls to register requirements such as "you need to have a mass (optionally: of value =...)". They should typically do so in their `__init__`.
- Requirements are given in form of a list containing 
    * Any directly instantiateable PyPIConGPU class (in practice, only `Constant`s and `Attribute`s are considered currently)
    * A `DelayedConstruction` of any PyPIConGPU class that cannot be instantiated directly. This typically happens when that class depends on a species itself.
    * A `DependsOn(species=...)` instance which (hopefully unsurprisingly) signals a dependency on another species (e.g. an ion species `DependsOn(species=electrons)` which they emit upon ionisation). 
- At the time of calling `Simulation.get_as_pypicongpu` all information is available by definition and a partial ordering is defined via `DependsOn` (cyclic dependencies are currently not checked). The translation can call `get_as_pypicongpu` on any entity, including the `Species` instances. This call is stateless, i.e., it can be repeated and the order of execution doesn't matter. Thus any entity can independently construct any PyPIConGPU class instances it needs.
- Afterwards, there's a global operation on all initialisation operations to fulfill some inter-species relationships that exist in the PyPIConGPU layer for no apparent reason.

As a side remark, one should note that it is undecided as of now if `register_requirements` should be considered public interface. It is designed as an implementation detail of our internal species handling. But it might provide an unintended resolution to some feature requests from @pordyna and @PrometheusPi. Their experience with the new protocol will show if the user experience is sufficiently intuitive to support its public usage. It is unlikely to make it into the PICMI standard.

For the documentation sprint:
- [ ] Write doc strings and documentation about the new species protocol.

## Dropping PyPIConGPU as an officially supported interface
I'm dropping PyPIConGPU as an officially supported interface. Its design goal was to be a Python representation of the input parameters of PIConGPU. It currently fails to meet this goal for no apparent reason. But it seems quite natural to move more into that direction. This would also resolve some of the above-mentioned unnecessary inter-species relationships. As we are also moving to Pydantic, this layer will literally contain no logic in itself, just a plain representation of data. In that state it will need no testing (beyond for typos which we can already test through PICMI). Once we have reached this state, it will be trivial to add this as an official interface again. For now, supporting, testing and thereby fixing its interface is a maintenance burden because there's thousands of lines of code that need adaption to the new interface. I have removed all PyPIConGPU unit tests. We are now free to change any behaviour unobservable from within PICMI.

In a future refactoring:
- [ ] Revisit the test cases and checks removed by this and re-implement their invariants as part of Pydantic validation.

## Other fixes and features
Some minor fixes, features, etc. sneaked in because they were on my to-do list for quite a while now:
- I have removed skipped unit tests in the PICMI layer. Those were tests for classes from the PICMI standard that we don't actually implement (yet).
- I have fixed the directory handling of the `pypicongpu.Runner`. It used to just `cd` around to its liking which internally messed up the working directory and made it hard to reason about where exactly something was executed (particularly while debugging). I have provided a context manager that makes sure to always return back to where it came from after changing directory for, e.g., running `pic-build`.
- Fixed a warning in `copy_attributes.py` because `inspect.getmembers` accesses `__fields__` which is deprecated in Pydantic v2. I'm now checking for `pydantic.BaseModel` and use the `model_fields` property instead in that case.
- Fix broken computation of typical ppc. This is a natural result of the refactoring I've done but it's worth noting because it will lead to generally larger values of the typical ppc.
- Add `method` attribute to `picmi.Species` according to PICMI standard. It indicates the pusher method used for this particle species. Our species class should now be standard-compliant.
- PICMI diagnostics are `BaseModel`s now.
- I have refrained from doing too much refactoring towards #5500 but I have performed some simplifications in that direction.
- Removed `Auto` diagnostic.

## Testing status
I have performed the following testing:
- The (remaining) `quick` and `compiling` tests pass (and will be checked in CI again).
- The `end-to-end` tests pass (and will not yet be checked in CI). Please run them again during review. They should run on any machine with a working PIConGPU environment including openPMD.
- I have diffed the generated setup from the [LWFA example](https://github.com/ComputationalRadiationPhysics/picongpu/tree/47cef4b3a0c943b1669b8dc21a9c092c76894684/share/picongpu/pypicongpu/examples/laser_wakefield) against its `mainline/dev` output. The result is identical up to whitespaces and the two known features/fixes (typical ppc and pusher method).

## Implementation details
The main algorithm of requirement resolution is implemented in `evaluate_requirements` in `picmi/species.py`. It basically filters for the type of requirement it's looking for, evaluates/resolves/??? that filtered collection and runs the delayed constructions. That intermediate "evaluates/resolves/???" step is a nasty, big, $\mathcal{O}(N^2)$ for-loop that on every pair of requirements
- checks for conflicts
- tries to merge them (for example, different forms of ionisation behaviour are bundled into one instance)
- drops unnecessary duplicates.

These steps are implemented as traits (free functions that run some default behaviour unless the given object(s) override that).

The `DelayedConstruction` is an interesting object, make sure you read its doc string. It is my solution to the conflict of requiring the freedom of arbitrary functions while at the same time having enough information/metadata about what is run to meaningfully reason about the intent, uniqueness, etc. that is encapsulated in that instance.

## For review
Just despair. There's no hope! It's thousands of lines of code and you'll never gonna make it!
...
Just kidding.

To the extent applicable to your situation:
- Please make sure that you run the `end-to-end` tests. (They are not yet run in CI.)
- Please run the examples.
- Please run any of your production cases and compare the output of `mainline/dev` with this branch. A convenient way of doing so for me was:
```bash
git checkout mainline/dev
# run picmi script with `write_input_file("expected")
git checkout -
# run picmi script with `write_input_file("new_branch")`
for file in $(ls {expected,new_branch}/include/picongpu/param/*.param | sed 's|^[^/]*/||g' | sort | uniq); do
    echo $file && diff {expected,new_branch}/$file;
done
```

I'd not primarily look at the new code in Github's diff view because most of it is a complete rewrite and there's little meaning to the diff. Start exploring it in your normal editor, jump around and get a feel for the code: Start from the `picmi.Simulation` class, specifically its `get_as_pypicongpu` method. Following the control flow will lead you through the algorithm of requirement resolution. Check the species internal requirements added from `picmi.Species.__init__` and the ionisation-related changes for examples of registering requirements. Afterwards you'll have covered the main functional code changes.

Back in the Github diff view, the remainder of the changes will hopefully feel intuitive and self-explanatory as part of one of the following categories:
- Necessary changes to adopt the new protocol
- Removal of the old protocol
- Removal and/or adaption of the tests
- Any item from the list of minor fixes and features

There's certainly a lot of bikeshedding applicable. Some namings and placements of code have not yet experienced the kind of careful consideration that they might deserve. Feel free to speak up about those. Some of the major ones I'm aware of:
- `_make_unique` in `species_requirements.py` is poorly named but I'm struggling to come up with a better name because it's quite an overload of responsibilities. That's probably a code smell but it's complex correlations being modeled here.
- There's surely a point to be made to move the provided specialisations of `DelayedConstruction` closer to their use cases. Currently, they are all bundled up in `species_requirements.py`. I was uncertain as to how much re-use there was to be expected among the code base when I started out, so I put them all into one central place. That will not scale well I'm afraid.
- `_DensityImpl` is poorly named and badly placed.

I have not cleaned up git history in any particular way and I'm uncertain what would be a safe and intuitive way to group these changes. It is admittedly quite a messy history but I'm afraid that the risk of introducing bugs by trying to disentangle and squash individual aspects into a small number of commits is quite high. And a single commit of ~8kloc of changes might also not be what we want.

chillenzer committed:
Introduce NEW1_Species

chillenzer committed:
Set up infrastructure

chillenzer committed:
Add new_species.py end-to-end test

chillenzer committed:
First species definition rendering

chillenzer committed:
Make speciesInitialization.param work

chillenzer committed:
Make particle.param pass

chillenzer committed:
Make density.param pass

chillenzer committed:
Switch test to iteration over filenames

chillenzer committed:
Add forgotten operation schema

chillenzer committed:
Update typical_ppc calculation

chillenzer committed:
Broaden Simulation constructor

chillenzer committed:
Extract method get_runner

chillenzer committed:
Remove some redundant comments

chillenzer committed:
Make tests run (update modern plugins)

chillenzer committed:
More robust directory handling in Runner

chillenzer committed:
Adapt other plugins to new species

chillenzer committed:
Strip picmi.Simulation from references to the init manager

chillenzer committed:
Ionisation introduces a dependency

chillenzer committed:
Make unique SetChargeState

chillenzer committed:
Reorder requirements

chillenzer committed:
Add merging and uniqueness protocol

chillenzer committed:
Introduce DelayedConstruction

chillenzer committed:
Fix Ionization stuff

chillenzer committed:
LWFA example works

chillenzer committed:
Add SimpleMomentum

chillenzer committed:
Purge auto diagnostics from codebase

chillenzer committed:
Remove reference to initmanager from docs

chillenzer committed:
Remove init manager and further details

chillenzer committed:
Fix RMS velocity handling

chillenzer committed:
Add mass and charge and handle constant conflicts

chillenzer committed:
Replace old Species with NEW1_Species

chillenzer committed:
Add shape and pusher support

chillenzer committed:
Move run_constructor responsibility higher up

chillenzer committed:
Merge LLKR4 and ReducedLandauLifshitz pusher entries

chillenzer committed:
Remove some more init_manager-related functions

chillenzer committed:
Make the tests run

chillenzer committed:
Minor renaming and refactoring

chillenzer committed:
Remove picmi.Species tests

chillenzer committed:
Infer species name from particle type if not given

chillenzer committed:
Remove picmi.Species tests (fix)

chillenzer committed:
Fix PICMI tests

chillenzer committed:
Adapt to new Species.__init__ interface

chillenzer committed:
More fixes

chillenzer committed:
Remove pypicongpu tests

chillenzer committed:
Remove skipped tests

chillenzer committed:
Remove pydantic/inspect deprecation warning

chillenzer committed:
Re-instantiate restriction of jobs for pypicongpu pic-build call

chillenzer committed:
Remove no-ops from pypicongpu

chillenzer committed:
Remove temporary new_species test

chillenzer committed:
Remove Auto diagnostic from warm_plasma example

chillenzer committed:
Fix test_minimal pypicongpu.Simulation test

chillenzer commented:
@PrometheusPi @pordyna @ikbuibui Should be ready for full review now. Please note (and read!) the updated description.

pordyna commented on lib/python/picongpu/picmi/species_requirements.py:
```diff
@@ -0,0 +1,273 @@
+        append_this = True
+        for rhs in result:
+            check_for_conflict(lhs, rhs)
+            merge_success = try_update_with(rhs, lhs)
```
What if `try_update_with` updates `rhs` into a state that changes the outcome of `check_for_conflict` that was already checked?

chillenzer replied:
Interesting question. Let's make an assumption to start off: `try_update_with` does not create new conflicts by itself, i.e., any conflict that the update-`rhs` has was already present in original-`rhs` or `lhs`. We can later revisit that assumption.

Control flow:
- Starting point: `result` contains a partial list and (the instance that in your question will be) `rhs` is supposed to be added to `result`.
- `rhs` checks for conflicts with everything already existing `result` and doesn't find any.
- `result` gets filled up further and every new entry will check for conflicts with the already added `rhs`.
- `lhs` is supposed to be added and checks for conflicts with everything in `result` up to `rhs` where it is merged.
- `result` is filled up to completion while every new addition checks against updated-`rhs`for conflicts.
- Finally, we arrive at `result = [..., updated-rhs, ..., <position where lhs would have been added>, ...]`.

This creates the following checking situation:
- `rhs` was checked against everything because
    - original-`rhs` was checked against everything left of would-be-`lhs` position.
    - updated-`rhs` was checked against everything right of updated-`rhs`.
    - We assumed that `try_update_with` doesn't create new conflicts.
- `lhs` was checked against everything left of updated-`rhs` and right of would-be-`lhs` position because
    - it was checked against everything left of updated-`rhs` before merging with `rhs`
    - updated-`rhs` was checked against everything right of would-be-`lhs` position.
    
This leaves a hole between updated-`rhs` and would-be-`lhs` position that `lhs` was never checked against. This hole could be closed by removing the short circuiting.

Is that assumption reasonable? We can just make it a requirement and revisit it if that's wrong. Checking for conflicts is anyhow more of a convenience feature. If there's conflicts it would typically blow up at some point. The main functionality of `check_for_conflict` is to fail early with a clear error message and not something cryptic about ambiguous linking or the like during compilation.

What are your thoughts about this?

chillenzer replied:
As we should hope for this not being performance critical, we could use the more concise multi-pass version
```python
def _make_unique(requirements):
    # multi-pass, so we need to make sure that requirements is stored persistently
    # we wouldn't want to exhaust the iterator on the first pass:
    requirements = list(requirements)

chillenzer replied:
Yes, I'm really not happy about this function. It's a difficult protocol.

chillenzer replied:
A different approach would be to maintain a list of possible actions, conflicts, etc. that we apply to the requirements:
```python
# Each of these returns a new list.
# The agreement is that they act on all pertinent entries
# and leave the rest as-is.
_RESOLUTION_ACTIONS = [
    # Checks for conflicts before dropping:
    DropDuplicateConstantsAndAttributes,
    MergeSimpleDensities,
    MergeIonizationModels,
    ...
]

# I'd probably rename this to resolve_requirements or the like
# but for the sake of consistency I'm sticking with _make_unique for now:
def _make_unique(requirements, actions=_RESOLUTION_ACTIONS):
    if len(actions) == 0:
        return requirements
    return _make_unique(actions[0](requirements), actions[1:])
```
Advantages:
- With this approach it is certainly easier to understand what actions are performed to resolve the requirements.
- The process is stateless./The style is functional which I like.

Disadvantages:
- Customisation of this pipeline would be centralised, i.e., if anybody wants to add a new action they'd have to hack something into this list. (Also, we've got a mutable default argument which is always dangerous.)
- I don't like relying on a global constant.
- One could instead hard-code the actions one-after-another into the function body but that's even less customisable.

chillenzer replied:
Further designs with the spirit of shifting responsibilities to an earlier stage:
- `Species.register_requirements` could already run the checking and the merging, etc. That would look similar to one of the above but it would be one loop each time a new element is registered instead of one big N^2 loop at the end. That doesn't sound like a significant gain but would already simplify things a little.
- We could expose the `Species._requirements` variable and allow registration via direct modification of that. So, the registering class is responsible for keeping requirements in a consistent state.
- `Species.register_requirements` could let the freshly registered requirement handle this:
```python
# trait to allow custom actions to be performed and provide some defaults otherwise:
def resolve(new_req, requirements):
    if hasattr(new_req, "resolve"):
        return new_req.resolve(requirements)
    if isinstance(new_req, Constant):
        if not is_unique_in(new_req, Constant):
            raise ValueError(...)
        return requirements + [new_req]
    ...

def register_requirements(self, new_requirements):
    for req in requirements:
        self.requirements = resolve(req, self.requirements)
```

chillenzer replied:
@pordyna @PrometheusPi You might want to add opinions on this design decision.

chillenzer replied:
@ikbuibui

chillenzer replied:
I'll look it up in the GoF again at home.

ikbuibui replied:
Ill take a closer look. As an aside for some interesting reading this problem reminds me of [this paper](https://sean-parent.stlab.cc/papers/2008-10-gpce/p89-jarvi.pdf).

ikbuibui replied:
Can the `check_for_conflict(lhs, rhs)` simply be moved into a separate loop after the merging is done with `try_update_with(rhs, lhs)`? Is the check before merging somehow different from after merging? Is one check at the end not sufficient?

pordyna commented on lib/python/picongpu/picmi/species.py:
```diff
@@ -5,328 +5,175 @@
+
+
+def particle_type_requirements(particle_type):
+    if (particle_type is None) or re.match(r"other:.*", particle_type):
```
What does the `other:` match do?

chillenzer replied:
That's [openPMD jargon](https://github.com/openPMD/openPMD-standard/blob/095799703baf69111e8383a6fda05ce029e65bb2/EXT_SpeciesType.md?plain=1#L31-L32) that [they've taken over into PICMI](https://github.com/picmi-standard/picmi/blob/e79dc49c58ad65b5e207ed40e4e6edc4b1482241/PICMI_Python/particles.py#L25-L27). Basically a wildcard to circumvent assumption about built-in particle types.

pordyna reviewed (commented):
Partial review, I just went over the files in my IDE, like suggested. Still need to look deeper into it.

PrometheusPi commented on lib/python/picongpu/picmi/interaction/ionization/ionizationmodel.py:
```diff
@@ -23,12 +26,38 @@ class IonizationModel(pydantic.BaseModel):
+            raise ValueError(f"{self.ion_species=} must be an ion.")
+        if self.ion_species.picongpu_fixed_charge:
+            raise ValueError(
+                f"I'm trying hard to ionize here but {self.ion_species.picongpu_fixed_charge=} is getting in the way."
```
(Optional) Do you really want to word this like this?

chillenzer replied:
I like it. You don't? ^^

PrometheusPi commented on lib/python/picongpu/picmi/__init__.py:
```diff
@@ -45,6 +45,7 @@
     "GaussianLaser",
     "PlaneWaveLaser",
     "Species",
+    "Species",
```
(Please fix) This is probably an error, `"Species"` is now loaded available twice.

chillenzer replied:
Done

PrometheusPi commented on lib/python/picongpu/picmi/copy_attributes.py:
```diff
@@ -89,7 +89,11 @@ def copy_attributes(
-        for from_name, _ in inspect.getmembers(from_instance)
+        for from_name, _ in (
+            type(from_instance).model_fields.items()
+            if isinstance(from_instance, BaseModel)
```
@chillenzer (Just a question) What is the advantage of distinguishing here between bot cases?

chillenzer replied:
This turns off a deprecation warning: `inspect.getmembers` tries to access `__fields__` which is deprecated in pydantic v2 in favour of the code you're looking at.

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/operation/setchargestate.py:
```diff
@@ -28,23 +26,18 @@ class SetChargeState(Operation):
 
-    def check_preconditions(self) -> None:
-        assert self.species.has_constant_of_type(GroundStateIonization), "BoundElectrons requires GroundStateIonization"
+    def __init__(self, species, charge_state):
```
(For me): `SetChargeState` constructor changed its interface.

PrometheusPi replied:
And since you started using "/" as positional-only parameter separator in [`simpledensity.py`](https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#diff-9f718c9b5dbf84bec0810cd34e482779dfa763fa219c6da46662853099bd5de1R46) why not here?

pordyna replied:
I think it did not change? The keywords are still 'species' and 'charge_state'

chillenzer replied:
Sorry, I got slightly confused due to having worked on #5583 already. This is a step towards #5500 (which will get finalised in #5583): While I'm not yet converting everything to pydantic `BaseModel`s yet, I'm already writing my new code such that it initialises everything at construction. In order to do so, I'm adding constructors where necessary. The `/` separator further aids the transition because pydantic will only accept explicitly named arguments in its constructor. It is thus optional but helpful to reduce the noise in #5583.

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/operation/simpledensity.py:
```diff
@@ -45,14 +41,14 @@ class SimpleDensity(DensityOperation):
-    def check_preconditions(self) -> None:
-        if self.ppc <= 0:
-            raise ValueError("must use positive number of particles per cell")
+    def __init__(self, /, species, profile, layout):
```
See [prior question](https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#r2631339097):
Why do you introduce a positional-only parameter separator here?

chillenzer replied:
As above: Helpful to enforce writing `SimpleDensity(species=...)` which is the form the construction will take when it's pydantic based.

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/operation/simpledensity.py:
```diff
@@ -122,7 +111,6 @@ def _get_serialized(self) -> dict:
             placed_species.append(species.get_rendering_context())
 
         return {
-            "ppc": self.ppc,
```
@chillenzer (but more a reminder for me): Where is ppc now stored/outputted?

chillenzer replied:
In the layout.

chillenzer replied:
PS: This was on my todo-list for a while now (roughly since #5371) because the density is not responsible for the cell-internal layout and therefore can't (and doesn't) necessarily know about ppc.

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/species.py:
```diff
@@ -33,6 +33,20 @@ class Shape(Enum):
 
 
+class Pusher(Enum):
+    # supported by standard and PIConGPU
```
(optional) 

```suggestion
    # supported by PICMI standard and PIConGPU
```

chillenzer replied:
Done

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/species.py:
```diff
@@ -33,6 +33,20 @@ class Shape(Enum):
+    Vay = "Vay"
+    Higuera = "Higuera-Cary"
+    Free = "Free"
+    # not supported by standard
```
(optional) 

```suggestion
    # only supported by PIConGPU
```

chillenzer replied:
Done

PrometheusPi commented on lib/python/picongpu/pypicongpu/species/species.py:
```diff
@@ -56,11 +70,20 @@ class Species(RenderedObject):
 
     shape = util.build_typesafe_property(Shape)
 
+    def __init__(self, /, name, constants=None, attributes=None, shape=None, pusher=None):
```
same question as [before](https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578#discussion_r2631339097)

chillenzer replied:
See above.

PrometheusPi commented on lib/python/picongpu/pypicongpu/runner.py:
```diff
@@ -306,8 +317,8 @@ def __render_templates(self):
         """launch build of PIConGPU"""
-        chdir(self.setup_dir)
-        runArgs("pic-build", ["pic-build", "-j", "4"])
+        with cd(self.setup_dir):
```
good catch. How did this work at all before?

chillenzer replied:
Well, depends what you mean by "work". The cwd wildly jumped around inside of the Python process and was never properly reset (but that didn't affect the calling process' working directory). You would only notice if you were to give relative paths for the different directories (don't remember if there was a conversion to absolute paths somewhere) or -- as was the case for me -- if you dropped into a breakpoint and found yourself totally lost in your filesystem.

PrometheusPi commented on share/picongpu/pypicongpu/template/etc/picongpu/N.cfg.mustache:
```diff
@@ -75,18 +75,6 @@ TBG_steps="{{{time_steps}}}"
-# only use charge conservation if solver is yee AND using cuda backend
-if [[ "Yee" = "{{{solver.name}}}" ]] && [[ "$PIC_BACKEND" =~ ^cuda(:.+)?$ ]]
-then
-    USED_CHARGE_CONSERVATION_FLAGS="--chargeConservation.period {{#data.period.specs}}{{{start}}}:{{{stop}}}:{{{step}}}{{^_last}},{{/_last}}{{/data.period.specs}}"
```
(Question) Why was this originally added?

chillenzer replied:
Admittedly: No idea! As it was inside of the `typeID.auto` tag, I felt confident to remove it without really having a look at what it was.

PrometheusPi commented on test/python/picongpu/quick/picmi/simulation.py:
```diff
@@ -146,12 +146,12 @@ def test_species_translation(self):
 
         # check typical ppc is derived
-        self.assertEqual(picongpu.typical_ppc, 2)
+        self.assertEqual(picongpu.typical_ppc, 3)
```
(Question/potential change): why was this change from 2 to 3?

chillenzer replied:
That was sort of a bug before (discussed in the internal chat already): The original algorithm for estimating `typical_ppc` considered a large class of operations some of which didn't even have a `ppc` value. Those were counted as contributing `1` or `0` or some mildly sensible default and thereby skewed the average heuristic applied here. As it's "only an estimate", it probably wasn't a critical bug but it didn't look intended in the code and led to some weird results. For example: There's only a single species with a layout of `ppc = 3` in this test but `typical_ppc` is 2.

PrometheusPi commented on test/python/picongpu/quick/picmi/simulation.py:
```diff
@@ -183,16 +183,15 @@ def test_wrong_explicitly_set_typical_ppc(self):
-            sim = picmi.Simulation(time_step_size=17, max_steps=4, solver=solver, picongpu_typical_ppc=value)
-            with self.assertRaisesRegex(ValueError, "typical_ppc must be >= 1"):
-                sim.get_as_pypicongpu()
+            with self.assertRaisesRegex(ValueError, "Typical ppc should be > 0"):
```
However, why was it changed in the first place?

PrometheusPi commented on test/python/picongpu/quick/picmi/simulation.py:
```diff
@@ -183,16 +183,15 @@ def test_wrong_explicitly_set_typical_ppc(self):
+            with self.assertRaisesRegex(ValueError, "Typical ppc should be > 0"):
+                picmi.Simulation(time_step_size=17, max_steps=4, solver=solver, picongpu_typical_ppc=value)
 
         wrongTypes = [0.0, -1.0, -15.0, 1.0, 15.0]
```
(essential question) ppc can/should(!) be float, why should this be  wrong type?
Kicking out negative values and zero is fine, but float should be allowed.

PrometheusPi commented on test/python/picongpu/quick/picmi/simulation.py:
```diff
@@ -285,30 +285,20 @@ def test_operations_simple_density_translated(self):
             if "separate1" in species_names or "colocated1" in species_names:
                 # used "layout"
-                self.assertEqual(3, op.ppc)
+                self.assertEqual(3, op.layout.ppc)
```
(confused question) here ppc is 3, as [above](https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#diff-7279559bfa16153e85397a461a178c41f935a572cc5734438a24c70a8134b633R154). Is this correlated?

chillenzer replied:
Yes, in the sense that the above is an average over this one value of 3.

PrometheusPi reviewed (changes requested):
There are some minor request for changes and several questions, some of them might indicate further changes. 

I did not yet review:
 - `lib/python/picongpu/pypicongpu/simulation.py`
 - `lib/python/picongpu/picmi/species_requirements.py`
 - `lib/python/picongpu/picmi/species.py`
 - `lib/python/picongpu/picmi/simulation.py`

and will do this (hopefully) tomorrow.

ikbuibui commented on lib/python/picongpu/picmi/species_requirements.py:
```diff
@@ -0,0 +1,273 @@
+            )
+
+        def try_update_with(self, other):
+            return (
+                isinstance(other, SimpleDensityOperation)
+                and other.metadata.kwargs["profile"] == self.metadata.kwargs["profile"]
+                and other.metadata.kwargs["layout"] == self.metadata.kwargs["layout"]
+                and (self.metadata.kwargs["species"].extend(other.metadata.kwargs["species"]) or True)
+            )
```
I really find this syntax confusing every time I see it

chillenzer replied:
It is admittedly a tiny bit code-golfy that there is a mutating operation in there. Would you be more comfortable with an explicit form?
```suggestion
            if (
                isinstance(other, SimpleDensityOperation)
                and other.metadata.kwargs["profile"] == self.metadata.kwargs["profile"]
                and other.metadata.kwargs["layout"] == self.metadata.kwargs["layout"]
                ):
                    self.metadata.kwargs["species"].extend(other.metadata.kwargs["species"])
                    return True
            return False
```

PrometheusPi replied:
@ikbuibui what is your opinion on @chillenzer's suggestion?

ikbuibui commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -168,6 +204,9 @@ def __init__(
 
+        if picongpu_typical_ppc is not None and picongpu_typical_ppc <= 0:
+            raise ValueError(f"Typical ppc should be > 0, not {self.picongpu_typical_ppc=}.")
+
```
Does the `PICMI_Simulation.__init__` modify `picongpu_typical_ppc`? else maybe the check should be called before the base class init

chillenzer replied:
No.

Fine with me. Do you have a particular reason?

chillenzer replied:
Done

ikbuibui commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -511,97 +341,118 @@ def _generate_openpmd_plugins(self, diagnostics, pypicongpu_by_picmi_species, nu
+
+def mid_window(iterable):
+    mi, ma = reduce(lambda lhs, rhs: (min(lhs[0], rhs), max(lhs[1], rhs)), iterable, (1000, 0))
+    return (ma - mi) // 2 + mi
```
What is the (1000,0)? and why the integer division?

chillenzer replied:
`(1000, 0)` is the neutral element of the `minmax` operation on the interval `[0, 1000]`. You might have follow-up question like: "Why on earth would you want to put 'the neutral element of the `minmax` operation on the interval `[0, 1000]`' in there?" To which I reply: "Okay, fine! I'll choose a less arbitrary range like `(-np.inf, np.inf)`." On a more serious note: This is currently used for typical ppc, so the values under consideration are on the order of 1-10 and they are all integer. The latter made me try not to reach for floats here.

Integer division because I want to have an integer. Maybe I should rename that function.

PrometheusPi replied:
@ikbuibui What is your opinion on @chillenzer's reply. 
And @chillenzer what is the status of your "Maybe I should rename that function."?

chillenzer replied:
I've removed the need for a neutral element by using the first element as the starting point.

I've made the function "_private" and added a small docstring clarifying the intent.

chillenzer commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -155,6 +188,9 @@ def __init__(
         **keyword_arguments,
     ):
+        self.species = []
+        self.layouts = []
```
Delete (will be set in `PICMI_Simulation.__init__` again).

chillenzer replied:
Done

chillenzer committed:
Reviewer comments

chillenzer committed:
Rename and reorganise resolve_requirements

chillenzer committed:
Add tests for resolve_requirements

chillenzer committed:
Minor tweak

chillenzer committed:
Add species requirement resolution tests

chillenzer committed:
Switch to resolving_add

chillenzer commented:
Switched to eager requirement resolution and added some not-at-all polished tests. We'll have to get back to this and properly reconsider the internal protocol. I think the existing code has sufficient encapsulation to not really affect other changes.

Check if you're happy now!

chillenzer committed:
Merge mainline/dev to resolve minor conflicts

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -511,97 +339,123 @@ def _generate_openpmd_plugins(self, diagnostics, pypicongpu_by_picmi_species, nu
 
-        s.base_density = self.picongpu_base_density or s.init_manager.get_base_density(s.grid)
+        init_operations = organise_init_operations(
+            chain(*(s.get_operation_requirements() for s in sorted(self.species)))
```
(question) Just for my understanding: why do you sort the `species`?

chillenzer replied:
The partial order defined on the species is the `DependsOn` relationship. If `species2` depends on `species1`, then `species1 < species2` and will be sorted accordingly. This becomes relevant in the generated C++ because that dependency will likely result in `species2`-related code referencing `species1` which must be declared before its first use. In an unrelated discussion (don't remember with whom), I've already clarified that I currently don't attempt to verify that it's really a well-defined partial order. In particular, I don't check for circles, etc.

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -511,97 +339,123 @@ def _generate_openpmd_plugins(self, diagnostics, pypicongpu_by_picmi_species, nu
+        moving_window = (
+            None
+            if self.picongpu_moving_window_move_point is None
+            else pypicongpu.movingwindow.MovingWindow(
```
(question) For my interest: Why do you prefer the ternary operator here instead of the old if-else block?

chillenzer replied:
That's a "declarative style". I think the main advantage is that under any circumstances, the following holds: Once I have passed this line, an assignment to `moving_window` has happened. There's nobody stopping by refactoring that code in a year such that
```python
if ...:
    moving_window = None
else:
    moving_window = ...
```
becomes
```python
if ...:
    # actually this condition also fits that other thing
    # so I'll just sneak in a little code here...
    moving_window = None
else:
    # of course there's also code coming in here
    # ...
    # Mmmhhh... just for testing I'll quickly disable that unrelated line
    # moving_window = ...
    # Ooops... now I dropped the originally intended statement in this block.
```

Also, there's a DRY argument to make because I'd have to write `moving_window = ` twice otherwise. That's a minor concern code-length-wise but could lead to refactorings missing to rename both, particularly because it's an assignment and it won't fire an exception due to "undefined name" or something like this.

chillenzer replied:
Also, it's semantically clearer: My version clearly states "I want to do an assignment to `moving_window`." (and that assignment happens to be non-trivial). The `if`-`else`-block version states "I want to do an operation in case of X and another operation otherwise. Oh, look! These two operations happen to be related!"

PrometheusPi commented on lib/python/picongpu/picmi/simulation.py:
```diff
@@ -511,97 +339,123 @@ def _generate_openpmd_plugins(self, diagnostics, pypicongpu_by_picmi_species, nu
-
-        return s
+    def _get_base_density(self) -> float:
+        # There's supposed to be some heuristics here along the lines of
```
commented out code
Is there a good reason to keep it (as it seems to document what needs to be done next)

chillenzer replied:
Mmmhhh... not sure. It's not "commented-out-instead-of-deleted" code. It's never been in use. It (and in fact the whole function) is the result of an early discussion, if we could heuristically determine `BASE_DENSITY` on the Python-level. In that sense, it serves a documenting purpose. I have other notes (and open tasks in those notes) on this topic, so it would hurt my personal productivity to remove it. Also, a Github issue might be a better place to preserve this as a to-do for future development.

I'd probably delete it in favour of an issue, if I have to touch this PR again.

chillenzer replied:
See #5598.

PrometheusPi replied:
fine with me

PrometheusPi commented on lib/python/picongpu/picmi/species.py:
```diff
@@ -1,332 +1,180 @@
+    # Theoretically, Position(), Momentum() and Weighting() are also requirements imposed from the outside,
+    # e.g., by the current deposition, pusher, ..., but these concepts are not separately modelled in PICMI
+    # particularly not as being applied to a particular species.
+    # For now, we add them to all species. Refinements might be necessary in the future.
```
good comment

PrometheusPi commented on lib/python/picongpu/picmi/species.py:
```diff
@@ -1,332 +1,180 @@
+        if self.name is None:
+            self.name = self.particle_type
+        try:
+            is_element = self.particle_type is not None and Element.is_element(self.particle_type)
```
why do you have to check first for `self.particle_type is not None` and then for `Element.is_element(self.particle_type)`? 
Wouldn't `Element.is_element(None)` not also return false?

PrometheusPi replied:
No: https://github.com/ComputationalRadiationPhysics/picongpu/blob/e8293d0414580ac80e504280d6fcc3ca12ee7a83/lib/python/picongpu/pypicongpu/species/util/element.py#L53
`Element.is_element` only accepts strings. 

Is there the possibility that `self.particle_type` is not a string or `None`?

PrometheusPi replied:
No: https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#diff-d2d07157a46143bb4f6f4ab0e5cabd432ea4622cc0a9f0824dea9052bf90beb7R58

PrometheusPi replied:
As this is in a `try` block, triggering a type-error via `Element.is_element(None)` would also go to the `except` path and set  `is_element` to `False` 

Thus the question remains: do you really need the extra `self.particle_type is not None` check?

chillenzer replied:
Well, as you said, it's a `TypeError` that I'd have to catch in addition. I we should revisit `is_element` at some point and make its interface more ergonomic. I wouldn't change anything here for now.

chillenzer replied:
PS: This code is more or less copied from the original implementation in the hope to minimize behavioural changes. I think this function should be re-organised in general (but in a separate PR).

PrometheusPi commented on lib/python/picongpu/picmi/species.py:
```diff
@@ -1,332 +1,180 @@
+            return None
+        try:
+            return (
+                pypicongpu.species.util.Element(self.particle_type) if Element.is_element(self.particle_type) else None
```
The only difference to https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#diff-d2d07157a46143bb4f6f4ab0e5cabd432ea4622cc0a9f0824dea9052bf90beb7R78 is that here None is returned not false or the Element if true. 

Fine with me, but looks like this is code duplication that could be avoided.

chillenzer replied:
That's a good observation. I'll use it in the model validator. Thanks!

PrometheusPi commented on lib/python/picongpu/picmi/species.py:
```diff
@@ -1,332 +1,180 @@
+
+    def __gt__(self, other):
+        # This defines a partial ordering on all species.
+        # This is necessary to determine the definition order inside of the C++ header.
```
Okay, that explains one of my previous questions of why you need to sort the species list.

PrometheusPi commented on lib/python/picongpu/picmi/species_requirements.py:
```diff
@@ -0,0 +1,279 @@
+            obj2.check_for_conflict(obj1)
+        if isinstance(obj1, Constant) and (isinstance(obj1, type(obj2)) or isinstance(obj2, type(obj1))):
+            if obj1 != obj2:
+                raise RequirementConflict(f"Conflicting constants {obj1=} and {obj2=} required.")
```
If I read this error message, I would not understand what is wrong. It sounds like, you require them to conflict. But you also catch that error later. 
What is going on here?

chillenzer replied:
Any of the `if` blocks can possibly raise. The peculiarity of the block you're pointing at lies solely in the fact that this is not hidden inside another method call.

The first exception raised here (which might be the one you're pointing at or an earlier `check_for_conflict`) will have relatively specific information like `RequirementConflict("Conflicting constants Mass(mass_si=1.) and Mass(mass_si=2.) required.")` or maybe (made-up) `ValueError("Ionization algorithm X can't be used in combination with property Y>5. You gave Y=7.)"`

I'm catching any exception here and I'm re-raising it as a `RequirementConflict`. The full error message will roughly look like this
```
Traceback (most recent call last):
[...]
RequirementConflict: Conflicting constants obj1=Mass(mass_si=1.) and obj2=Mass(mass_si=2.) required.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
[...]
RequirementConflict: A conflict in requirements between obj1=Mass(mass_si=1.) and obj2=Mass(mass_si=2.) has been detected. See above error message for details.
```

I will change "Conflicting constants..." to "Conflicting values for constants..." which is hopefully a little bit clearer in that specific case.

I'll totally give you that for the conflicting values of a constant this seems to be a bit of an overkill. But I hope the `ValueError` example exhibits the full benefit:
```
Traceback (most recent call last):
[...]
ValueError: Ionization algorithm X can't be used in combination with property Y>5. You gave Y=7.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
[...]
RequirementConflict: A conflict in requirements between obj1=IonizationX(...) and obj2=PropertyY(...) has been detected. See above error message for details.
```
It's a tiny bit unfortunate that I don't have information about which species I'm concerned with at this point. But we'll handle that if it ever turns out to be a problem.

PrometheusPi replied:
Thanks for the explanation. I think this is much clearer will all printed error messages. Only the last one (without context) was confusing.

PrometheusPi commented on lib/python/picongpu/picmi/species_requirements.py:
```diff
@@ -0,0 +1,279 @@
+def must_be_unique(requirement):
+    return (hasattr(requirement, "must_be_unique") and requirement.must_be_unique) or (
+        isinstance(requirement, Constant) or isinstance(requirement, Attribute)
+    )
```
Comment for me: This returns either `requirement.must_be_unique` if attribute exists or the bool value of the later or calls. But this is fine as `must_be_unique` has to be a bool too. 

Resolved.

PrometheusPi commented on lib/python/picongpu/picmi/species_requirements.py:
```diff
@@ -0,0 +1,279 @@
+        def constructor(self):
+            species = self.metadata.kwargs["species"].get_as_pypicongpu()
+            particle_mass_si = species.get_constant_by_type(Mass).mass_si
+            rms_velocity_si_squared = np.linalg.norm(self.metadata.kwargs["rms_velocity"]) ** 2 / 3
```
Why do you move the division by 3 here? 
This would be per component. 
Could you add this to the name or do the division later?

PrometheusPi commented on test/python/picongpu/quick/picmi/species.py:
```diff
@@ -1,381 +1,91 @@
+
+def subset_of(subset, superset):
+    superset = list(superset)
+    return all(any(e1 == e2 for e2 in superset) for e1 in subset)
```
I think we had this discussion last year, why `set(subset).issubset(superset)` does not work here, right? Could you please remind me, what the issue was (also for documentation)?

chillenzer replied:
"hashable" is the keyword. Content of sets (and the keys of dictionaries) must be hashable and I don't want that constraint (and, in fact, I couldn't satisfy it).

PrometheusPi commented on test/python/picongpu/quick/picmi/species.py:
```diff
@@ -1,381 +1,91 @@
+
+def count_all_one(subset, superset):
+    superset = list(superset)
+    return all(superset.count(e) == 1 for e in subset)
```
What is the difference to `subset_of`?

PrometheusPi replied:
as far as I can see it has the same result.

PrometheusPi replied:
`count_all_one([1,2,2], [1,2,3,4])` returns `True` despite having `2` twice.

chillenzer replied:
Sorry, I see how that's bad naming. It is supposed to mean "Is the count of all elements in X exactly one in Y?", so
`count_all_one(X=[1], Y=[1,2,2])` is `True` but `count_all_one(X=[1,2], Y=[1,2,2])` is `False`. Phrased differently it's `subset_of(X,Y) and are_all_unique_in([y for y in Y if y in X])`. I'll come up with a better name. Suggestions?

chillenzer replied:
I've renamed the function `unique_in` and moved the `all` to the outer scope. So, now it's
```
unique_in([1,2], [1,2,2]) == [True, False]
all(unique_in([1,2], [1,2,2])) == False
```

PrometheusPi commented on test/python/picongpu/quick/picmi/species.py:
```diff
@@ -1,381 +1,91 @@
+        # Ionization makes the ion depend on the electron species.
+        # This is important for rendering the corresponding C++ header,
+        # so the electron species gets defined before the ion species.
+        assert electron < ion
```
This is cool 😄

chillenzer replied:
Yes, I've totally taken a particle physics perspective here: Internally, the code connects to the LHC, sets up an appropriate experiment and determines the size of both objects. It takes a few decades to run for every execution because it first has to get the grant application to pass and hire the personnel.

PrometheusPi commented on test/python/picongpu/quick/picmi/species.py:
```diff
@@ -1,381 +1,91 @@
+        ]
+        # They have been merged:
+        assert len(ground_state_ionizations) == 1
+        assert len(ground_state_ionizations[0].ionization_model_list) == len(ionizations)
```
Could you please explain what this does?

chillenzer replied:
On the PICMI-level we have constructed multiple different ionisation algorithm instances that all ionise the `ion` and produce the `electron`. These instances have internally registered their requirements to the `ion` species.

On the PIConGPU level, the generated code will look schematically like this:
```
ionizers<MakeSeq_t<Ionization1, Ionization2, ...>>
```
So, the PyPIConGPU representation has all of them in one object which is a `Constant` of the (translated) `ion` species.

What I want to check is (schematically)
```python
filter(is_ionization, ion.get_as_pypicongpu().constants) == get_as_pypicongpu(ionizations)
```
The ionizations should be translated correctly. This is not quite trivial to achieve, so I'm only using some proxies here:
- There should be exactly one `GroundStateIonization` in the `ion` requirements because they have all been merged.
- This `GroundStateIonization` should contain exactly one entry for each ionization I've set up.

That's weaker but significantly easier to check and provided sufficient coverage at that time.

PrometheusPi replied:
Just a follow-up question unrelated to the PR (not blocking): Do you enforce an order of ionizer algorithms? E.g. it makes sense to run BSI before ADK as it cheaper and covers ADK's issue with classical ionization.

chillenzer replied:
No, there's currently no such code. One could easily do the same trick that I've done with the species, i.e., define a partial order on all ionizations and add a `self...model_list'] = sorted(self...model_list'])` [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/5578/changes#diff-029143c764322adbedb1cb9050a76a4881c1dcd5dc93dd577c8b2d4eb63f9c18R212). But I haven't yet looked into what those actually do, so I wouldn't know which order to apply.

PrometheusPi commented on lib/python/picongpu/picmi/species_requirements.py:
This file is beyond my understanding.

chillenzer replied:
=(
What can we do to make this simpler?

PrometheusPi replied:
I am afraid not. Have to trust you on this one or @ikbuibui or @pordyna have to have another review.

PrometheusPi reviewed (changes requested):
Some questions that need clarification. Looks otherwiese good.

chillenzer committed:
Reviewer comments
